File: doma-core/src/main/java/org/seasar/doma/jdbc/criteria/declaration/ValuesDeclaration.java
/*
 * Copyright Doma Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.seasar.doma.jdbc.criteria.declaration;

import java.util.Objects;
import org.seasar.doma.jdbc.criteria.context.InsertContext;
import org.seasar.doma.jdbc.criteria.context.Operand;
import org.seasar.doma.jdbc.criteria.metamodel.PropertyMetamodel;

public class ValuesDeclaration {

  private final InsertContext context;

  public ValuesDeclaration(InsertContext context) {
    this.context = Objects.requireNonNull(context);
  }

  public <PROPERTY> void value(PropertyMetamodel<PROPERTY> left, PROPERTY right) {
    Objects.requireNonNull(left);
    context.values.put(new Operand.Prop(left), new Operand.Param(left, right));
  }
}

File: doma-criteria/src/main/kotlin/org/seasar/doma/criteria/declaration/ValuesDeclaration.kt
package org.seasar.doma.criteria.declaration

import org.seasar.doma.criteria.context.Operand
import org.seasar.doma.jdbc.Config

@Declaration
class ValuesDeclaration(val config: Config) {
    private val expressionDeclaration = ExpressionDeclaration(config)

    fun expression(block: ExpressionDeclaration.() -> Operand.Expr): Operand.Expr {
        return expressionDeclaration.block()
    }
}
